resource-validator/src/main/scala/com/broadinstitute/dsp/resourceValidator/DataprocWorkerChecker.scala

```diff
@@ -0,0 +1,67 @@
+package com.broadinstitute.dsp.resourceValidator
+
+import cats.effect.{Concurrent, Timer}
+import cats.mtl.Ask
+import cats.implicits._
+import com.broadinstitute.dsp.{
+  regionName,
+  resourceValidator,
+  CheckRunner,
+  CheckRunnerConfigs,
+  CheckRunnerDeps,
+  RuntimeCheckerDeps,
+  RuntimeWithWorkers
+}
+import io.chrisdavenport.log4cats.Logger
+import org.broadinstitute.dsde.workbench.google2.DataprocClusterName
+import org.broadinstitute.dsde.workbench.model.TraceId
+
+object DataprocWorkerChecker {
+  def impl[F[_]: Timer](
+    dbReader: DbReader[F],
+    deps: RuntimeCheckerDeps[F]
+  )(implicit F: Concurrent[F], logger: Logger[F], ev: Ask[F, TraceId]): CheckRunner[F, RuntimeWithWorkers] =
+    new CheckRunner[F, RuntimeWithWorkers] {
+      override def appName: String = resourceValidator.appName
+
+      override def configs = CheckRunnerConfigs(s"dataproc-workers", shouldAlert = true)
+
+      override def dependencies: CheckRunnerDeps[F] = deps.checkRunnerDeps
+
+      override def resourceToScan: fs2.Stream[F, RuntimeWithWorkers] = dbReader.getRuntimesWithWorkers
+
+      override def checkResource(runtime: RuntimeWithWorkers, isDryRun: Boolean)(
+        implicit ev: Ask[F, TraceId]
+      ): F[Option[RuntimeWithWorkers]] =
+        for {
+          clusterOpt <- deps.dataprocService
+            .getCluster(runtime.r.googleProject, regionName, DataprocClusterName(runtime.r.runtimeName))
+          runtime <- clusterOpt.fold[F[Option[RuntimeWithWorkers]]](F.pure(None)) { c =>
+            val doesPrimaryWorkerMatch =
+              runtime.workerConfig.numberOfWorkers == c.getConfig.getWorkerConfig.getNumInstances
+            val doesSecondaryWorkerMatch =
+              runtime.workerConfig.numberOfPreemptibleWorkers == c.getConfig.getSecondaryWorkerConfig.getNumInstances
+            val isAnomalyDetected = !(doesPrimaryWorkerMatch && doesSecondaryWorkerMatch)
+
+            isAnomalyDetected match {
+              case true =>
+                val log =
+                  if (isDryRun)
+                    logger
+                      .warn(
+                        s"${runtime} has an anomaly with the number of workers in google. \n\tPrimary work match status: ${doesPrimaryWorkerMatch}\n\tSecondary worker match status: ${doesSecondaryWorkerMatch}"
+                      )
+                  else
+                    // we want to see if this occurs before doing anything. Initial scans show 0 occurrences in dev.
+                    logger.warn(
+                      s"${runtime} has an anomaly with the number of workers in google. \n\tPrimary work match status: ${doesPrimaryWorkerMatch}\n\tSecondary worker match status: ${doesSecondaryWorkerMatch}"
```

Comment:
shall we resize the cluster if not in `dryRun` mode?

Reply:
I don't think we should  do to all the limitations around worker number changing  (i.e. if its the primary workers that don't match, sometimes the cluster needs to be deleted/re-created. This seems problematic if a user is working). I think we should see if this actually happens in the wild before investing the time to do this, as it would most likely require a pubsub message and adding the delete -> update -> re-create flow to leo similar to the stop -> update -. start flow we have. Even then, it could  be surprising for users.

I think the best course of action is to see if it actually happens, and then we can do the easy re-size of secondary  workers and layer in the primary worker update messaging

Reply:
@kyuksel since you're the expert on resizing dataproc cluster now, what do you think?

Reply:
> I don't think we should do to all the limitations around worker number changing (i.e. if its the primary workers that don't match, sometimes the cluster needs to be deleted/re-created.

I wasn't aware of this limitation since for my work, I've only been concerned with resizing to remove preemptibles. I didn't see such a limitation be mentioned in the [docs](https://cloud.google.com/dataproc/docs/concepts/configuring-clusters/scaling-clusters#:~:text=After%20creating%20a%20Dataproc%20cluster,Use%20Dataproc%20Autoscaling.) either. Let me play around quickly with it and report back.

> I think we should see if this actually happens in the wild before investing the time to do this

I can see how that can make sense. What would be the checkpoint mechanism to re-evaluate where things stand on this? The concern being we may forget...

Also a tangential question: I wonder if we should only take into consideration Dataproc clusters in `Running` status, especially if we decide to take action on anomalies.